lib/relooper/src/graph/supergraph.rs

```diff
@@ -0,0 +1,340 @@
+                .contained
+                .iter()
+                .map(|&l| (l, *self.cfg.edge(&l)))
+                .filter(|(_l, e)| e.to_vec().into_iter().any(|&to| to == split_snode.head))
+                .collect();
+
+            for (f, e) in from_split {
+                let redirected = e.redirect(|to| *versions_mapping.get(&to).unwrap_or(&to));
+                self.cfg.remove_edge(f, e);
+                self.cfg.add_edge(f, redirected);
+            }
+
+            // populate supernode graph with new node's new version (for each split)
+            let splitted_head = versions_mapping.get(&split_snode.head).unwrap().to_owned();
+            let contained: Vec<_> = versions_mapping.values().copied().collect();
+            self.nodes.insert(
+                splitted_head,
+                SNode::new(
+                    splitted_head,
+                    BTreeSet::from_iter(contained.iter().copied()),
+                ),
+            );
+            let node_ref = self.nodes.get(&splitted_head).unwrap();
+            for c in contained {
+                self.label_location.insert(c, node_ref.head);
+            }
+        }
+    }
+
+    /// iterates over supergraph (using `snode_order` on each iteration) until there is only one node left
+    /// on each iteration we either:
+    /// * remove one supernode (by merging it into another one)
+    /// * or duplicate one (by splitting, one dup for each "parent")
+    /// in the end, there is only one supernode, which contains all the nodes and whose head is "entry" node
+    pub fn reduce(&mut self) {
+        let mut in_edges: Option<HashMap<SLabel<TLabel>, HashSet<SLabel<TLabel>>>> = None;
+        'outer: loop {
+            let order: Vec<SLabel<TLabel>> = self.snode_order();
+
+            let mut splits: HashMap<SLabel<TLabel>, SplitInto<TLabel>> = Default::default();
+
+            for snode_label in order {
+                let n = self.nodes.get(&snode_label).unwrap();
+                let in_edges = in_edges.get_or_insert_with(|| self.cfg.in_edges());
+                match self.node_action(n, in_edges) {
+                    None => {}
+                    Some(SplitFor(split)) => {
+                        splits.insert(n.head, split);
+                    }
+                    Some(MergeInto(to)) => {
+                        self.merge(n.head, to);
+                        continue 'outer;
+                    }
+                }
+            }
+
+            let mut split_len: BTreeMap<usize, Vec<&SLabel<TLabel>>> = BTreeMap::new();
+
+            for (split_node, split_for) in splits.iter() {
+                split_len
+                    .entry(split_for.len())
+                    .or_default()
+                    .push(split_node);
+            }
+
+            if let Some((_, biggest_splits)) = split_len.last_key_value() {
+                // let split_internal_nodes = biggest_splits.iter().map(|(n, split_for)|); // TODO select by internal node count?
```

Comment:
Please remove commented out code if it is not needed.

Reply:
I think it's fixed

Reply:
Thanks for deleting the commented out code. But the todo remains, will something be done about it or should it be turned into a GitHub issue?

Reply:
Hm, I think this function contains quite big and interesting task that can be an issue. 

There is a lot of reducible graphs that are equivalent to given irreducible, but they are different in the size. We want generate new graph as small as possible. In the paper about relooper was given a greedy algorithm, Pavel implemented some more complicated algorithm. But we don't have proofs if that algorithm is best possible, and I think it is not the best. So, I think some time in the future it will be good to make research about this algorithms and implement the best one.